standard/expressions.md

```diff
@@ -2370,6 +2371,77 @@ A *default_value_expression* is a constant expression ([§12.20](expressions.md#
 ### 12.7.16 Anonymous method expressions
 
 An *anonymous_method_expression* is one of two ways of defining an anonymous function. These are further described in [§12.16](expressions.md#1216-anonymous-function-expressions).
+	
+### 12.7(.x) §stack-allocation Stack allocation
+
+A stack allocation initializer allocates a block of memory from the call stack.
+
+```ANTLR
+stackalloc_initializer
+    : 'stackalloc' unmanaged_type '[' expression ']'
+    | 'stackalloc' unmanaged_type? '[' expression? ']' stackalloc_initializer_elements
+    ;
+
+stackalloc_initializer_elements
+    : '{' stackalloc_initializer_element_list? '}'
+    | '{' stackalloc_initializer_element_list ',' '}'
+    ;
+
+stackalloc_initializer_element_list
+    : stackalloc_element_initializer (',' stackalloc_element_initializer)*
+    ;
+    
+stackalloc_element_initializer
+    : expression
+    ;
+```
+
+The *unmanaged_type* ([§9.8](types.md#98-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.
+
+If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer_elements*. If *expression* is omitted from *stackalloc_initializer*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer_elements*.
+
+When a *stackalloc_initializer* includes both *expression* and *stackalloc_initializer_elements* the number of elements in that *stackalloc_initializer_elements* shall match the value of *expression*.
```

Comment:
Should *expression* be required to be a constant expression in that case, so that the compiler can check that the length matches? Thus making the following invalid even if `i` would be 1 at run time.

```C#
using System;
public class C {
    public void M(int i) {
        Span<byte> s = stackalloc byte[i] {1};
    }
}
```

Reply:
[12.7.11.5 Array creation expressions](https://github.com/dotnet/csharpstandard/blob/cf0e08d15a437a1773e6da213358de6fba1a7ee6/standard/expressions.md#127115-array-creation-expressions) states a similar requirement:

<https://github.com/dotnet/csharpstandard/blob/cf0e08d15a437a1773e6da213358de6fba1a7ee6/standard/expressions.md#L1924>